File: JavaScript/huffmanDecoder.js
const encodeString = require('./huffmanEncoder');

//const huffmanObj = encodeString('Data compression using Huffman Coding, currently completed an encoder in JavaScript. Capable of compressing small strings, not optimized for paragraphs or files yet. I would like to continue to grow this repository to support many languages and to be able to compress large text files and more.');


//Going to switch to searching prefix table instead for optimization when decoding, 
//reference python repo for example
function traversal(string, tree) {
	let refTree = tree;
	let arr = string.split('');
	let result = '';
	let index = 0;
	while(arr.length) {
		if(!refTree.left && !refTree.right) {
			if(refTree.data[0] === '■') {
				break;
			} else {
				result+=refTree.data[0];
				process.stdout.write(refTree.data[0]);
				refTree = tree;
				arr = arr.slice(index);
				index = 0;
			}
		}
		if(arr[index] === '0') {
			refTree = refTree.left;
		} else {
			refTree = refTree.right;
		}
		index++;
	}
	console.log(' ');
}

function decode(obj) {
	const tree = obj.key;
	const data = obj.huffmanString;
	traversal(data, tree);
}

//decode(huffmanObj);
module.exports = traversal;

File: Python/HuffmanDecoder.py
from HuffmanEncoder import encodeString, generatePrefixes

#took about 5 seconds with current 16KB file
def traversal(string, tree):
	ref = tree
	arr = list(string)
	result = ''
	index = 0
	while len(arr) > 0:
		if ref.left is None and ref.right is None:
			if ref.data[0] == '■':
				break
			else:
				result += ref.data[0]
				ref = tree
				arr = arr[index:]
				index = 0
		if arr[index] == '0':
			ref = ref.left
		else:
			ref = ref.right
		index += 1
	return(result)

def prefixArray(tree):
	prefixes = generatePrefixes(tree)
	arr = []
	for obj in prefixes:
		arr.append([obj, prefixes[obj]])
	arr.sort(key=lambda x: len(x[1]))
	return arr


#takes 0.3 seconds with current 16KB
#takes 1.64 seconds for 140KB
#takes 16.54 seconds for 1.4MB
def decodeFromArray(tree, string):
	prefixArr = prefixArray(tree)
	result = ''
	temp = ''
	index = 0
	for char in string:
		temp += char
		for obj in prefixArr:
			if temp == obj[1]:
				if obj[0] == '■': break
				result += obj[0]
				temp = ''
				break
	print(result)
